File: internal/service/webhook/webhookService.go
package webhook

import (
	"bytes"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"strings"
	"time"

	"prospectsync-server/internal/models"
)

type Session struct {
	baseURL    string
	secret     string
	event      models.WebhookEvent
	typ        models.WebhookType
	fullURL    string
	httpClient *http.Client
}

func Initiate(event models.WebhookEvent, eventType models.WebhookType) (*Session, error) {
	base := strings.TrimSuffix(os.Getenv("WEBHOOK_BASE_URL"), string(event))
	secret := os.Getenv("WEBHOOK_SECRET")

	if base == "" || secret == "" {
		return nil, fmt.Errorf("webhook config is missing: base=%q secret=%q", base, secret)
	}

	fullURL := fmt.Sprintf("%s/%s", base, event)
	return &Session{
		baseURL:    base,
		secret:     secret,
		event:      event,
		typ:        eventType,
		fullURL:    fullURL,
		httpClient: &http.Client{Timeout: 10 * time.Second},
	}, nil
}

func (s *Session) send(payload *models.WebhookPayload) error {
	payload.Event = s.event
	payload.Type = s.typ
	payload.Timestamp = time.Now().UTC().Format(time.RFC3339Nano) // date is a ref for GO to format timestamp

	body, err := json.Marshal(payload)

	log.Printf("⭐ DEBUG webhook POST -> %s\nPayload bytes: %d\nPayload JSON: %s\n", s.fullURL, len(body), string(body))
	if err != nil {
		return fmt.Errorf("json: %w", err)
	}

	mac := hmac.New(sha256.New, []byte(s.secret))
	mac.Write(body)
	signature := "sha256=" + hex.EncodeToString(mac.Sum(nil))

	log.Printf("⭐ DEBUG webhook signature: %s\n", signature)

	req, _ := http.NewRequest("POST", s.fullURL, bytes.NewBuffer(body))
	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-webhook-signature", signature)

	resp, err := s.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode >= 400 {
		msg, _ := io.ReadAll(resp.Body)
		log.Printf("DEBUG webhook response (%d): %s", resp.StatusCode, string(msg))
		return fmt.Errorf("webhook %d: %s", resp.StatusCode, string(msg))
	}

	return nil
}

func (s *Session) Start(msg ...string) error {
	payload := &models.WebhookPayload{
		Event:  s.event,
		Type:   s.typ,
		Status: models.StatusStarted,
	}

	if len(msg) > 0 {
		payload.Data = map[string]any{"msg": msg[0]}
	}

	return s.send(payload)
}

func (s *Session) Running() error {
	return s.send(&models.WebhookPayload{
		Event:  s.event,
		Type:   s.typ,
		Status: models.StatusRunning,
	})
}

func (s *Session) Success(data any) error {
	return s.send(&models.WebhookPayload{
		Event:  s.event,
		Type:   s.typ,
		Status: models.StatusSuccess,
		Data:   data,
	})
}

func (s *Session) Error(msg string) error {
	return s.send(&models.WebhookPayload{
		Event:  s.event,
		Type:   s.typ,
		Status: models.StatusError,
		Error:  msg,
	})
}

File: src/services/webhook/webhook.service.ts
import crypto from "crypto";
import { env } from "../../config/env";
import { WebhookPayload, WebhookEvent, WebhookType, WebhookStatus } from "../../models/webhook.model";

export class WebhookService {
  private baseURL: string;
  private secret: string;

  constructor() {
    this.baseURL = env.WEBHOOK_BASE_URL;
    this.secret = env.WEBHOOK_SECRET;
  }

  initiate(event: WebhookEvent, eventType: WebhookType, payloadOwner?: string): WebhookSession {
    const base = this.baseURL.endsWith(event) ? this.baseURL.slice(0, -event.length) : this.baseURL;
    const fullURL = `${base}${event}`;
    console.log("🔗 URL: ", fullURL);

    if (!base || !this.secret) {
      throw new Error(`webhook config is missing: base=${base} secret=${this.secret ? "***" : ""}`);
    }

    return new WebhookSession(fullURL, this.secret, event, eventType, payloadOwner);
  }
}

export class WebhookSession {
  constructor(
    private fullURL: string,
    private secret: string,
    private event: WebhookEvent,
    private typ: WebhookType,
    private payloadOwner?: string
  ) {}

  private async send(payload: WebhookPayload): Promise<void> {
    payload.event = this.event;
    payload.type = this.typ;
    payload.timestamp = new Date().toISOString();
    if (this.payloadOwner) {
      payload.payloadOwner = this.payloadOwner;
    }

    const body = JSON.stringify(payload);

    // Create HMAC signature
    const mac = crypto.createHmac("sha256", this.secret);
    mac.update(body);
    const signature = `sha256=${mac.digest("hex")}`;

    const response = await fetch(this.fullURL, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-webhook-signature": signature,
      },
      body,
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`webhook ${response.status}: ${errorText}`);
    }
  }

  async start(msg?: string): Promise<void> {
    const payload: WebhookPayload = {
      event: this.event,
      type: this.typ,
      status: WebhookStatus.STARTED,
      timestamp: new Date().toISOString(),
    };

    if (msg) {
      payload.data = { msg };
    }

    return this.send(payload);
  }

  async running(): Promise<void> {
    return this.send({
      event: this.event,
      type: this.typ,
      status: WebhookStatus.RUNNING,
      timestamp: new Date().toISOString(),
    });
  }

  async success(data: any, msg?: string): Promise<void> {
    return this.send({
      event: this.event,
      type: this.typ,
      status: WebhookStatus.SUCCESS,
      data,
      message: msg,
      timestamp: new Date().toISOString(),
    });
  }

  async error(msg: string): Promise<void> {
    return this.send({
      event: this.event,
      type: this.typ,
      status: WebhookStatus.ERROR,
      error: msg,
      timestamp: new Date().toISOString(),
    });
  }
}

let instance: WebhookService | null = null;

export function getWebhookService(): WebhookService {
  if (!instance) instance = new WebhookService();
  return instance;
}
